### PR TITLE
RSE-1089: Fix Security Loophole with Case.GetList API

### PR DESCRIPTION
**Pull request by tunbola**
## Overview
The Case.GetList API allows to pass a parameter called `params`, this parameter can be used to pass a list of case categories that the user does not have access to and it will return results. This PR fixes the issue.

## Before
The Issue described above exists.

## After
The Case.getList API can now be called without passing any case category related parameters and it will return results filtered by the case categories that the logged in contact has access to. When the case categories are passed, there is a check to make sure that the user has access to all the case categories passed otherwise an exception is thrown.

## Technical Details
- The `CRM_Civicase_Hook_alterAPIPermissions_CaseGetList` class was added to add permissions for the Case.getList API and also the CaseContact.get API based on if the user has access to any case categories. The permission was added also for CaseContact.get because this API is also called by Case.getList and permission is checked. See [here](https://github.com/civicrm/civicrm-core/blob/master/api/v3/Case.php#L754-L759) and [here](https://github.com/civicrm/civicrm-core/blob/master/api/v3/Generic/Getlist.php#L40)
- The `CRM_Civicase_Api_Wrapper_CaseGetList` class is added to ensure that the case type category parameter is validated when passed and if not present it is added and limited to the case type categories that the user has access to.
- The Case.getList API calls some other API's behind the scenes that require permission access, one of such is the Case.get API because it needs to join to the API to get some information. The function for fetching case category to allow set permission for this API was modified to support when the case type category is passed using the `IN` operator.
- Only the `=` operator and the `IN` operator is supported for passing case type categories. It makes sense because if this check is not it place, the API can be spoofed and also we can't check for all the operators.
- The permission check for the `basic case information` permission in `civicase_civicrm_selectWhereClause` is removed and also causing issues and it is is un-necessary because the entry API has already checked the needed permission to access the API. Also another way to maintain this is to loop through and check for equivalent case category permission but it is un-necessary.

## Comments
- An API wrapper may be needed for the CaseContact.get API later to also limit calls to case type categories but that will  done in another task as it is outside the scope of this task.

### CODE REVIEW

File: CRM/Civicase/Api/Wrapper/CaseGetList.php [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+  /**
+   * {@inheritdoc}
+   */
+  public function toApiOutput($apiRequest, $result) {
+
+    return $result;
+  }
+
+  /**
+   * Checks whether the request can be handled or not.
```
Comment (reneolivo) [inline CRM/Civicase/Api/Wrapper/CaseGetList.php:41]: Can we describe when the request can be handled instead? Like we only accept requests coming from the `Case.getlist` endpoint.

File: CRM/Civicase/Api/Wrapper/CaseGetList.php [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    }
+
+    $accessibleCaseCategories = CaseCategoryHelper::getAccessibleCaseTypeCategories();
+    $caseCategoryParameterValue = $this->getCaseTypeCategoryParameterValue($apiRequest['params']);
+
+    if (!empty($caseCategoryParameterValue)) {
+      $this->validateCaseTypeCategoryParameter($caseCategoryParameterValue, $accessibleCaseCategories);
+    }
+    else {
+      $apiRequest['params']['params']['case_id.case_type_id.case_type_category'] = ['IN' => $accessibleCaseCategories];
```
Comment (reneolivo) [inline CRM/Civicase/Api/Wrapper/CaseGetList.php:26]: can we move the `['IN'....` part to a new line? This line is too long to be able to read:

```php
$apiRequest['params']['params']['case_id.case_type_id.case_type_category'] = [
  'IN' => $accessibleCaseCategories,
];
```

File: CRM/Civicase/Api/Wrapper/CaseList.php [hunk 3]
```diff
@@ -1,149 +1,153 @@
...
+            unset($result['values'][$key]);
+          }
+        }
+      }
+    }
+
+    return $result;
+  }
+
+}
```
Comment (reneolivo) [inline CRM/Civicase/Api/Wrapper/CaseList.php:153]: 🍰 Just a small comment for next time: When moving files from one place to another it's better to make every change in their own commit. In this case we should have fixed lint issues first and then moved the file. The reason for this is to make it easier to understand what changed.

File: ang/civicase/activity/actions/services/move-copy-activity-action.service.js [hunk 4]
```diff
@@ -69,19 +69,7 @@
...
-
-      return {
-        'case_type_id.case_type_category': firstCaseTypeCategoryId,
-        params: {
-          'case_id.case_type_id.case_type_category': {
-            IN: allCaseTypeCategoryNames
-          }
-        }
-      };
+      return {};
```
Comment (reneolivo) [inline ang/civicase/activity/actions/services/move-copy-activity-action.service.js:72]: I think this change might break some tests. Also, I think we don't need the `getCategoriesWithAccessToBasicCaseInformation` function anymore, do we @deb1990 ?

File: api/v3/Case/Getcaselist.php [hunk 5]
```diff
@@ -1,48 +1,55 @@
...
+ *   Parameters to be passed to API call to obtain case list.
+ *
+ * @return array
+ *   API result with the list of cases.
+ */
+function civicrm_api3_case_getcaselist(array $params) {
+  $caseList = new CRM_Civicase_Api_Wrapper_CaseList();
+
+  return $caseList->getCaseList($params);
+}
```
Comment (reneolivo) [inline api/v3/Case/Getcaselist.php:55]: Same comment as 🍰

File: CRM/Civicase/Api/Wrapper/CaseGetList.php [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+use CRM_Case_BAO_CaseType as CaseType;
+use CRM_Civicase_Helper_CaseCategory as CaseCategoryHelper;
+
+/**
+ * CRM_Civicase_ApiWrapper_CaseGetList.
+ */
+class CRM_Civicase_Api_Wrapper_CaseGetList implements API_Wrapper {
+
+  /**
+   * {@inheritdoc}
```
[the remark this answers is not in the document]
Comment (tunbola) [reply CRM/Civicase/Api/Wrapper/CaseGetList.php]: Done.

File: CRM/Civicase/Api/Wrapper/CaseGetList.php [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+    }
+    else {
+      $apiRequest['params']['params']['case_id.case_type_id.case_type_category'] = ['IN' => $accessibleCaseCategories];
+    }
+
+    return $apiRequest;
+  }
+
+  /**
+   * {@inheritdoc}
```
[the remark this answers is not in the document]
Comment (tunbola) [reply CRM/Civicase/Api/Wrapper/CaseGetList.php:40]: This function does not do any custom implementation so I'll leave as it is

**Comment by reneolivo** [comment]
Tests results:

#### Admin User
✅ `{}`: returns a mix of cases and awards.
✅ `{params: { 'case_id.case_type_id.case_type_category': {IN: ['awards']}}}`: returns awards only.
✅ `{params: { 'case_id.case_type_id.case_type_category': {IN: ['Cases']}}}`: returns cases only.
✅ `{params: { 'case_id.case_type_id.case_type_category': {IN: ['awards', 'Cases']}}}`. returns a mix of cases and awards.
✅ `{params: { 'case_id.case_type_id.case_type_category': 'awards'}}`: returns awards only.
✅ `{params: { 'case_id.case_type_id.case_type_category': 'Cases'}}`: returns cases only.

#### Case User
✅ `{}`: Returns cases only.
✅ `{params: { 'case_id.case_type_id.case_type_category': {IN: ['awards']}}}`: Throws an error `"User does not have access to this case category awards"`.
✅ `{params: { 'case_id.case_type_id.case_type_category': {IN: ['Cases']}}}`: returns cases only.
✅ `{params: { 'case_id.case_type_id.case_type_category': {IN: ['awards', 'Cases']}}}`: Throws an error `"User does not have access to this case category awards"`.
✅ `{params: { 'case_id.case_type_id.case_type_category': 'awards'}}`: Throws an error `"User does not have access to this case category awards"`.
✅ `{params: { 'case_id.case_type_id.case_type_category': 'Cases'}}`: returns cases only.
